### ISSUE TITLE
Document coding style (and perhaps formatting tool options)

### ISSUE THREAD

**Issue by rhansen** [ddclient/ddclient#129]
It would be nice to document code style such as tabs vs. spaces, indentation level, tightness of parentheses, etc. in a contributor guide.

Perhaps the guide should encourage the use of a code formatter. I used the following [perltidy](https://metacpan.org/pod/perltidy) command to help with PR #120:

    perltidy -l=0 -conv -ci=4 -ola -ce -nbbc -kis -pt=2 -o=ddclient.new ddclient

**Comment by wimpunk** [comment]
I know, the style is a pretty mess. Feel free to document.

**Comment by SuperSandro2000** [comment]
How do we want to define URL builder?

Do we want to use a verbose variant?

```perl
$url  = "http://$config{$h}{'server'}/api/autodns.cfm";
$url .= "?id=";
$url .= $h;
$url .= "&pw=";
$url .= $config{$h}{'password'};
$url .= "&ip=";
$url .= $ip;
$url .= "&client=";
$url .= $config{$h}{'client'};
```

a shorter variant?

```perl
$url  = $globals{'ssl'} ? "https://" : "http://";
$url .= $config{$h}{'server'} . $config{$h}{'script'};
$url .= "?username=$config{$h}{'login'}";
$url .= "&password=$config{$h}{'password'}";
$url .= "&ip=$ip";
$url .= "&id=$h";
```

or a super short variant?

```perl
my $url = "http://$config{$h}{'server'}$config{$h}{'script'}?system=";
```

I would personally prefer the shorter variant and change all URL builders to closely resemble this.

**Comment by SuperSandro2000** [comment]
Next question: How do we want to format the example here docs? Right now they are not very uniform and I think it would help people if they where formatted in a uniform style and order.

Below are some of my personal opinions I would like to suggest:

I copied examples from several places together but this generally applies to all example here docs.

```
  protocol=nsupdate \\
  server=ns1.example.com \\
  password=/etc/${program}/dyn.example.com.key \\
  zone=dyn.example.com \\
  ttl=3600 \\
  password=myg3nerat3dp4ssword      \\
  myhost.dyn.example.com
```

This example has the escapes right after the variable assignment which is unique. Most other examples have them at the end with an line length of about 60 to 66.

I would suggest that we enforce a line length of 64 and put a single escape character at the end. The double escape is wrong but I am going to fix this in a PR.

This example is also missing commas which we should enforce to better handle multi domain updates by default.

Passwords right now are random and chosen by the contributor. They range from good examples (``your_password``) to very long ones (``123456789ABCDEF0000000000000000000000000000000000000``). I would suggest that we use a single password in the entire documentation except if some API key is required.

For the password I suggest ``your-password`` or ``service-password``. For API keys I would use a format that combines the service name with the word ``-api-key`` combined. An example would be ``dondominio-api-key``.

The same is true for the username. I suggest we stick to ``service-username``.

Then I would loosely follow the updated order. If some provider has different variables we try to match them in a order that is least service/user specific to most user specific.

The updated example would like:

```
  protocol=nsupdate,                                           \
  server=ns1.example.com,                                      \
  password=/etc/${program}/dyn.example.com.key,                \
  zone=dyn.example.com,                                        \
  ttl=3600,                                                    \
  password=myg3nerat3dp4ssword,                                \
  myhost.dyn.example.com
```

Then there are explanation blocks like:

```
  protocol=dyndns2             ##
  server=fqdn.of.service       ## defaults to members.dyndns.org
  script=/path/to/script       ## defaults to /nic/update
  backupmx=no|yes              ## indicates that this host is the primary MX for the domain.
  static=no|yes                ## indicates that this host has a static IP address.
  custom=no|yes                ## indicates that this host is a 'custom' top-level domain name.
  mx=any.host.domain           ## a host MX'ing for this host definition.
  wildcard=no|yes              ## add a DNS wildcard CNAME record that points to {host}
  login=service-login          ## login name and password  registered with the service
  password=service-password    ##
  fully.qualified.host         ## the host registered with the service.
```

Most of them have 31 characters before the ``##``. I would like to keep this and make it a rule. If some option needs to be longer a single line can have more characters before the ``##``. The explanation can have more than one line but they should not be more than two.

I would leave the short intro text as a free form field. It should mention what service this update function is for and the website.

I am not sure why there is an ``o`` before each name. I would remove them as the end user is probably as much confused as I am right now.

A final example could than look like this:

```perl
sub nic_dtdns_examples {
    return <<'EoEXAMPLE';
o 'dtdns'
The 'dtdns' protocol is the protocol used by the dynamic hostname services
of the 'DtDNS' dns services. This is currently used by the free
dynamic DNS service offered by www.dtdns.com.
Configuration variables applicable to the 'dtdns' protocol are:
  protocol=dtdns               ##
  server=www.fqdn.of.service   ## defaults to www.dtdns.com
  password=service-password    ## password registered with the service
  client=name-of-updater       ## defaults to $program (10 chars max, no spaces)
  fully.qualified.host         ## the host registered with the service.
Example ${program}.conf file entries:
  ## single host update
  protocol=dtdns,                                              \
  password=service-password,                                   \
  client=ddclient,                                             \
  example.dtdns.net
EoEXAMPLE
}
```


We also have a lot of separators (``######################################################################``= in our code. They are right now 70 chars long and I would leave them like this.

Some points I am unsure about:

- Do we want to keep ``server=www.fqdn.of.service`` or change it to the example domain ``server=www.example.org``?
- Do we want to stick with the example update domain ``fully.qualified.host`` or change it to some subdomain of the example domain ``sub.example.com``?
- Is the variable in ``Example ${program}.conf file entries:`` supposed to be replaced? If so I would just replace it with a static ``ddlcient.conf`` cause it is broken right now anyway.

I would love to hear @wimpunk @DaveSophoServices and @rhansen opinion about this and I would start to change all the examples to be more uniform after I wrote this quick style guide down in our CONTRIBUTING.md together with the perltidy formatting options.


PS Sorry for the unordered long text but I think a consistent documentation would help new users a lot to better understand what they need to set and what it should be.

**Comment by rhansen** [comment]
> How do we want to define URL builder?

My preference would be to use the [URI module](https://metacpan.org/pod/URI) to build the URLs. That will take care of escaping special characters in query parameters.

> How do we want to format the example here docs?

Backslash continuations don't mix well with comments, so I think it would be best to encourage config files that look like this:

```
key1=value1
key2=value2
key3=value3
hostname1, hostname2, hostname3
```

Technically this sets `key1`, `key2`, and `key3` globally (and the globals are copied into the per-host settings once the hostnames line is read), but that subtlety is unimportant for most use cases.

> This example is also missing commas which we should enforce to better handle multi domain updates by default.

Could you explain how the commas are important? Looking at the code they're optional except for separating hostnames.

> I would suggest that we use a single password in the entire documentation except if some API key is required.

:+1:

> For the password I suggest `your-password` or `service-password`.

I have a slight preference for `your-password`, but it doesn't really matter to me.

> For API keys I would use a format that combines the service name with the word `-api-key` combined. An example would be `dondominio-api-key`.

:+1: 

> The same is true for the username. I suggest we stick to `service-username`.

I have a slight preference for `your-username`, but it doesn't really matter to me.

> I am not sure why there is an `o` before each name.

Probably intended to be a bullet point.

> We also have a lot of separators in our code.

I think those are largely noise and would prefer to delete them, but I'm OK with them staying.

>   * Do we want to keep `server=www.fqdn.of.service` or change it to the example domain `server=www.example.org`?
> 
>   * Do we want to stick with the example update domain `fully.qualified.host` or change it to some subdomain of the example domain `sub.example.com`?

`www.example.org` is too generic. Is that referring to the user's host or the service provider's host?

How about `your-host.example.com` for the user's host and `service-provider.example` for the service provider's host?

>   * Is the variable in `Example ${program}.conf file entries:` supposed to be replaced?

Yes.

> If so I would just replace it with a static `ddlcient.conf` cause it is broken right now anyway.

:+1:

As an aside, I think that URL building style and example formatting style are too high-level to mention in a style guide. Once we unify the existing styles, most future contributions will naturally copy that style unless there's a good reason not to.

**Comment by SuperSandro2000** [comment]
> My preference would be to use the URI module to build the URLs. That will take care of escaping special characters in query parameters.

This sounds like we should open a new issue for that.

> Technically this sets key1, key2, and key3 globally (and the globals are copied into the per-host settings once the hostnames line is read), but that subtlety is unimportant for most use cases.

This needs to be documented.

> Could you explain how the commas are important? Looking at the code they're optional except for separating hostnames.

I did not look at the code I must admit. Lets drop them if they are not needed.

> I think those are largely noise and would prefer to delete them, but I'm OK with them staying.

I didn't want to introduce to much changes for the introduction. If the others also think they are noise we just remove them.

> How about your-host.example.com for the user's host and service-provider.example for the service provider's host?

👍 

> As an aside, I think that URL building style and example formatting style are too high-level to mention in a style guide. Once we unify the existing styles, most future contributions will naturally copy that style unless there's a good reason not to.

You are probably right but to unify the current style we need a common ground at least internally.

**Comment by rhansen** [comment]
I think this has been sufficiently addressed by #144.
